test/simulator/utils.cpp

```diff
@@ -0,0 +1,30 @@
+#include "utils.hpp"
+
+#include "logger.hpp"
+#include "simulator.hpp"
+
+namespace test {
+
+void add_links(sim::Simulator& sim,
+               std::initializer_list<two_way_link_t> links) {
+    for (auto& l : links) {
+        sim.add_link(l.first, l.second, 0, 0);
+        sim.add_link(l.second, l.first, 0, 0);
+    }
+}
+
+bool check_reachability(std::shared_ptr<sim::IRoutingDevice> src_device,
+                       std::shared_ptr<sim::IRoutingDevice> dest_device) {
+    auto curr_device =
+        src_device->get_link_to_destination(dest_device)->get_to();
+    while (curr_device != dest_device) {
+        if (curr_device == nullptr) {
+            return false;
+        }
+        curr_device =
+            curr_device->get_link_to_destination(dest_device)->get_to();
+    }
+    return true;
+}
```

Comment:
Incorrect routing table may contain loops that will lead to infinite `while`. To deal with it, mark devices you meet. Also, please, be careful with nullptr
```suggestion
static std::shared_ptr<sim::IRoutingDevice> > get_next_device(std::shared_ptr<sim::IRoutingDevice> curr_device,
                                                                                                     std::shared_ptr<sim::IRoutingDevice> dest_device) {
    auto next_link = curr_device->get_link_to_destination(dest_device);
    if (link == nullptr) {
        return nullptr;
    }
    return link->get_to();
}

bool check_reachability(std::shared_ptr<sim::IRoutingDevice> src_device,
                       std::shared_ptr<sim::IRoutingDevice> dest_device) {
    std::set<std::shared_ptr<sim::IRoutingDevice> > used;
    auto curr_device = src_device;
    while (curr_device != dest_device) {
        if (curr_device == nullptr || used.contains(curr_device) ) {
            return false;
        }
        used.insert(curr_device);
        curr_device = get_next_device(curr_device, dest_device);
    }
    return true;
}
```